<a href="https://colab.research.google.com/github/ranieri-unimi/git.ammagamma/blob/main/WITHOUT_solution_TEXT_CLASSIFICATION.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Boring stuff: setting everything up

*Warning: run this section only once*

Connect to your Google Drive so that your work does not get lost when you end your session

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


Change working directory to your Google Drive

In [ ]:
%cd /content/drive/MyDrive/

/content/drive/MyDrive


Create the main directory for the laboratory inside your Google Drive

In [ ]:
!mkdir NLP_MASTER

mkdir: cannot create directory ‘NLP_MASTER’: File exists


Remove unwanted directories (if it is your first run these directories do not exist and the following two commands have no effect)

In [ ]:
!rm -rf /content/drive/MyDrive/NLP_MASTER/finance

In [ ]:
!rm -rf /content/drive/MyDrive/NLP_MASTER/spacy-projects

Now let's install all the dependencies for the laboratory

In [ ]:
!pip install -U spacy-nightly --pre

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 12.7 MB 5.1 MB/s 
     |████████████████████████████████| 9.1 MB 32.5 MB/s 
     |████████████████████████████████| 660 kB 63.8 MB/s 
     |████████████████████████████████| 42 kB 1.8 MB/s 
     |████████████████████████████████| 457 kB 68.9 MB/s 
     |████████████████████████████████| 58 kB 8.0 MB/s 
  Attempting uninstall: typing-extensions
    Found existing installation: typing-extensions 4.2.0
    Uninstalling typing-extensions-4.2.0:
      Successfully uninstalled typing-extensions-4.2.0
  Attempting uninstall: catalogue
    Found existing installation: catalogue 1.0.0
    Uninstalling catalogue-1.0.0:
      Successfully uninstalled catalogue-1.0.0
  Attempting uninstall: srsly
    Found existing installation: srsly 1.0.5
    Uninstalling srsly-1.0.5:
      Successfully uninstalled srsly-1.0.5
  Attempting uninstall: smart-open
    Found exis

In [ ]:
!pip install -U pip setuptools wheel

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 2.1 MB 5.3 MB/s 
     |████████████████████████████████| 1.2 MB 56.2 MB/s 
  Attempting uninstall: setuptools
    Found existing installation: setuptools 57.4.0
    Uninstalling setuptools-57.4.0:
      Successfully uninstalled setuptools-57.4.0
  Attempting uninstall: pip
    Found existing installation: pip 21.1.3
    Uninstalling pip-21.1.3:
      Successfully uninstalled pip-21.1.3
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
spacy 2.2.4 requires catalogue<1.1.0,>=0.0.7, but you have catalogue 2.0.7 which is incompatible.
spacy 2.2.4 requires srsly<1.1.0,>=1.0.2, but you have srsly 2.4.3 which is incompatible.
spacy 2.2.4 requires thinc==7.4.0, but you have thinc 8.0.17 which is incompatible.
datascience 0.10.6 requir

In [ ]:
!pip install -U spacy transformers

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.2/6.2 MB 53.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.2/4.2 MB 59.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 181.6/181.6 kB 24.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 596.3/596.3 kB 48.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.6/6.6 MB 47.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 86.2/86.2 kB 10.8 MB/s eta 0:00:00
  Attempting uninstall: pyyaml
    Found existing installation: PyYAML 3.13
    Uninstalling PyYAML-3.13:
      Successfully uninstalled PyYAML-3.13
  Attempting uninstall: spacy
    Found existing installation: spacy 2.2.4
    Uninstalling spacy-2.2.4:
      Successfully uninstalled spacy-2.2.4


Now that everything is set up, change working directory to the newly created directory NLP_MASTER in your Google Drive

In [ ]:
%cd /content/drive/MyDrive/NLP_MASTER/

/content/drive/MyDrive/NLP_MASTER


Cline the official projects from the Spacy Repo, you are going to start from [this one](https://github.com/explosion/projects/tree/v3/tutorials/textcat_goemotions) and adapt it to the sentiment classification of financial news headlines

In [ ]:
!git clone https://github.com/explosion/projects.git spacy-projects

Cloning into 'spacy-projects'...
remote: Enumerating objects: 3742, done.
remote: Counting objects: 100% (85/85), done.
remote: Compressing objects: 100% (51/51), done.
remote: Total 3742 (delta 38), reused 73 (delta 32), pack-reused 3657
Receiving objects: 100% (3742/3742), 13.48 MiB | 9.61 MiB/s, done.
Resolving deltas: 100% (2293/2293), done.
Checking out files: 100% (410/410), done.


Let's now create a subdirectory "finance" inside NLP_MASTER, where we are going to copy the textcat_goemotions tutorial we just cloned with git with the command above

In [ ]:
!mkdir finance

In [ ]:
!cp -r spacy-projects/tutorials/textcat_goemotions/* finance/

In [ ]:
%cd /content/drive/MyDrive/NLP_MASTER/finance/

/content/drive/MyDrive/NLP_MASTER/finance


Spacy command line in action: now that we moved in the root directory of the project we tell Spacy to download everything the project needs in order to be run

In [ ]:
!spacy project assets

ℹ Fetching 4 asset(s)
✔ Downloaded asset
/content/drive/MyDrive/NLP_MASTER/finance/assets/categories.txt
✔ Downloaded asset
/content/drive/MyDrive/NLP_MASTER/finance/assets/train.tsv
✔ Downloaded asset
/content/drive/MyDrive/NLP_MASTER/finance/assets/dev.tsv
✔ Downloaded asset
/content/drive/MyDrive/NLP_MASTER/finance/assets/test.tsv


# Sentiment analysis: Reddit Posts Dataset

*Example records [TEXT_CONTENT, EMOTION_ID, TEXT_ID]:*

You can take a look at the dataset [here](https://drive.google.com/file/d/118kEBuOXikDJhlAvDVmAVxNBymtQ5MKb/view?usp=sharing)

*   My favourite food is anything I didn't have to cook myself.	27	eebbqej 
*   Thank you friend	15	eeqd04y
*   It's crazy how far Photoshop has come. Underwater bridges?!! NEVER!!!	7,13	efanc6t


Check out **assets/categories.txt** to explore the labels for this dataset. *The first row corresponds to the emotion_id 0, the second row to the emotion_id 1 and so on.*

---



##***Edit [project.yml](/content/drive/MyDrive/NLP_MASTER/finance/project.yml) and change gpu_id from -1 to 0 in order to take advantage of the Colab GPU***

Let Spacy **preprocess Reddit Posts Dataset** (assets/train.tsv, assets/dev.tsv, assets/test.tsv and assets/categories.txt) and format it as it internally needs.

In [ ]:
!spacy project run preprocess


================================= preprocess =================================
ℹ Skipping 'preprocess': nothing changed


Now that the dataset has been processed, **let's train the model** on the Reddit posts!

In [ ]:
!spacy project run train


=================================== train ===================================
Running command: /usr/bin/python3 -m spacy train ./configs/cnn.cfg -o training/cnn --gpu-id 0
ℹ Saving to output directory: training/cnn
ℹ Using GPU: 0

=========================== Initializing pipeline ===========================
[2022-06-08 11:32:35,839] [INFO] Set up nlp object from config
[2022-06-08 11:32:35,847] [INFO] Pipeline: ['textcat']
[2022-06-08 11:32:35,851] [INFO] Created vocabulary
[2022-06-08 11:32:35,852] [INFO] Finished initializing nlp object
[2022-06-08 11:33:00,965] [INFO] Initialized pipeline components: ['textcat']
✔ Initialized pipeline

============================= Training pipeline =============================
ℹ Pipeline: ['textcat']
ℹ Initial learn rate: 0.001
E    #       LOSS TEXTCAT  CATS_SCORE  SCORE 
---  ------  ------------  ----------  ------
  0       0          0.30       50.18    0.50
  0     200          8.00       55.12    0.55
  0     400          7.09       58.45 

Automatic SpaCy evaluation of the model you just trained

In [ ]:
!spacy project run evaluate


================================== evaluate ==================================
Running command: /usr/bin/python3 -m spacy evaluate ./training/cnn/model-best ./corpus/test.spacy --output ./metrics/cnn.json
ℹ Using CPU
ℹ To switch to GPU 0, use the option: --gpu-id 0

================================== Results ==================================

TOK                   100.00
TEXTCAT (macro AUC)   83.49 
SPEED                 28793 


=========================== Textcat F (per label) ===========================

                      P       R       F
admiration        64.56   60.71   62.58
amusement         76.61   85.61   80.86
anger             56.04   25.76   35.29
annoyance         46.38   10.00   16.45
approval          53.85   15.95   24.62
caring            54.84   12.59   20.48
confusion         44.90   14.38   21.78
curiosity         52.38   30.99   38.94
desire            51.35   22.89   31.67
disappointment    66.67    1.32    2.60
disapproval       40.00   14.98   21.80
disgu

Let's test the model on some examples, **feel free to change them to whatever you want**!

In [ ]:
import spacy
nlp = spacy.load("./training/cnn/model-best")

texts = [
    "It was really bad to watch you leave, hopefully you'll be back soon",
    "Oh yes, I can relate to that. Still, you'd better think about it twice.",
]

for doc in nlp.pipe(texts):
    # Do something with the doc here
    print(doc.cats)

{'admiration': 0.001072859624400735, 'amusement': 0.005732352379709482, 'anger': 0.007456898223608732, 'annoyance': 0.00208781729452312, 'approval': 0.0018803244456648827, 'caring': 0.04452180489897728, 'confusion': 0.0008455381612293422, 'curiosity': 0.003570307046175003, 'desire': 0.0202154740691185, 'disappointment': 0.0325545109808445, 'disapproval': 0.004504743497818708, 'disgust': 0.05633332580327988, 'embarrassment': 0.006672602612525225, 'excitement': 0.0031072725541889668, 'fear': 0.0030973341781646013, 'gratitude': 0.0028433476109057665, 'grief': 0.006908386945724487, 'joy': 0.003084561787545681, 'love': 0.002204897813498974, 'nervousness': 0.005881026852875948, 'optimism': 0.9313751459121704, 'pride': 0.0013122900854796171, 'realization': 0.0041063628159463406, 'relief': 0.005566315725445747, 'remorse': 0.015642797574400902, 'sadness': 0.1405908167362213, 'surprise': 0.0007496478501707315, 'neutral': 0.0026976754888892174}
{'admiration': 0.07745881378650665, 'amusement': 0.0

#Data Preparation: from the Reddit Post Dataset to the Financial News Dataset
**TODO: Upload Financial News Dataset file FinancialPhraseBank_AllAgree.txt to the assets folder, you can find the dataset [here](https://drive.google.com/file/d/1WXM2t8sh-myIEUZt37zIXC2McNrCyS2l/view?usp=sharing)**\

Financial news dataset example records [TEXT_CONTENT, SENTIMENT_LABEL]:


*   According to Gran , the company has no plans to move all production to Russia , although that is where the company is growing .@neutral
*   Finnish Talentum reports its operating profit increased to EUR 20.5 mn in 2005 from EUR 9.3 mn in 2004 , and net sales totaled EUR 103.3 mn , up from EUR 96.4 mn .@positive
*   Pharmaceuticals group Orion Corp reported a fall in its third-quarter earnings that were hit by larger expenditures on R&D and marketing .@negative



---

Now you have to **format the Financial News Dataset like the Reddit Posts Dataset**, in order to retrain the sentiment classifier on the new financial dataset.

Remember to split the dataset into train (70%), validation (10%) and test (20%), **saving the respective TSV files (train.tsv, dev.tsv, test.tsv) in the asset folder** .

---
## Hints:
- Our final dataset should have the following columns: text, label, id. Text and label are already in our file (in the same row!), while the ID should be generated uniquely (e.g. use uuid.uuid4())
- Categories are represented as strings (neutral, positive, negative), while spacy expects them as integer.
- Should we split the observations randomly or use some specific criteria?
- The train, val and test files should be stored as tab separated files (sep="\t") under the assets/ folder, with the following names: 
  - train.tsv
  - dev.tsv
  - test.tsv

In [ ]:
### TO DO

If you didn't do it before, check out the file under assets/categories.txt : it contains the (many) labels for the sentiment classification of the Reddit Posts Dataset, now you have to **change it to the labels of the Financial News Dataset (neutral, positive, negative)**.

In [ ]:
#!echo -en "neutral\npositive\nnegative" > /content/drive/MyDrive/NLP_MASTER/finance/assets/categories.txt
!echo -en "neutral\npositive\nnegative" > ./assets/categories.txt

Let again Spacy **preprocess our input files** (assets/train.tsv, assets/dev.tsv, assets/test.tsv and assets/categories.txt) and format them as it internally needs.

In [ ]:
!spacy project run preprocess

Spacy is a bit picky about existing directories, **delete the previous CNN model** you trained on the Reddit Posts Dataset

In [ ]:
#!rm -rf /content/drive/MyDrive/NLP_MASTER/finance/training/cnn
!rm -rf ./training/cnn

Everything is ready, **let's train the model** on the Financial News Dataset!

In [ ]:
!spacy project run train

In [ ]:
!spacy project run evaluate

# Running predictions on examples!

In [ ]:
import spacy
nlp = spacy.load("./training/cnn/model-best")

texts = [
    "Net income was $9.4 million compared to the prior year of $2.7 million.",
    "Revenue exceeded twelve billion dollars, with a loss of $1b.",
]

for doc in nlp.pipe(texts):
    # Do something with the doc here
    print(doc.cats)


# Final task: sentiment as a Prophet regressor

## Main goal
**The presentations will start at 11:45 a.m. on Friday**
Forecast the EUR-USD exchange rate, using both timeseries (e.g. previous values) and news Downloaded from [ForexRate news archive.](http://www.forexrate.co.uk/newsarchive.php). 
## Dataset
You will use the dataset downloaded and used in previous labs.
As a test set use observations in the range [1st June 2021, 1st June 2022], extremes included.

In your presentation you should focus on the methodological approach you used for solving this problem **AND** the main insights to share with your business stakeholders.

## Metrics
You should use some of the metrics shown during the time-series lecture (or even better ones!) and motivate your choices. It will be certainly interesting to go beyond stating "the MAE is X.Y": are there any particular patterns? how performances varies throughout time? is it worth having a predictive model instead of "baseline" approaches?
## Presentation format
Each team, made of 3/4 members, will present their results to all of us in 15 minutes, using a brief Power point presentation and answer eventual questions (both from us and other teams!).

## Organizational stuff
**The presentations will start at 11:45 a.m. on Friday**

Until then you can work together with your team mates: please don't work on it overnight!!!
For us it's more interesting to see which insights will you share with business stakeholders and the statistical robustness of your methodological approaches, instead of seeing an infinitesimal improvement on your metrics of choice.

We will be available also tomorrow morning, from 9 a.m., for answering all your questions and/or help you solve some technical issues on the dedicated call.




In [ ]:
import requests
from bs4 import BeautifulSoup

offset = 0
max_offset = 1649
offset_increment = 12

BASE_URL = 'http://www.forexrate.co.uk/'

In [ ]:
news_archive = []

for i in range(0,max_offset,offset_increment):
  url = f'http://www.forexrate.co.uk/newsarchive.php?start={i}'
  print(url)
  page = requests.get(url)
  soup = BeautifulSoup(page.content, 'html.parser')
  tables = soup.findChildren('table')
  news_table = tables[1]
  rows = news_table.findChildren(['th', 'tr'])

  for idx,row in enumerate(rows):
    if idx == 0:
        continue
    cells = row.findChildren('td')
    for idx,cell in enumerate(cells):
      txt = cell.text
      href = cell.find('a')['href']
      href = BASE_URL + href.replace('./','')
      if "newsarchive.php?start=" in href:
        continue
      # let's get the date of the article
      date_page = requests.get(href)
      date_soup = BeautifulSoup(date_page.content, 'html.parser')
      date_div = date_soup.findChildren('div')[3]
      date_str = date_div.text
      news_archive.append({'txt':txt,'url':href,'date':date_str})
      print(len(news_archive), date_str, {'txt':txt,'url':href,'date':date_str})
      #print(value, href)
  #print(len(news_archive))

#print(news_archive)

In [ ]:
import pandas as pd

df = pd.DataFrame(news_archive)

In [ ]:
df.to_csv("./hist_fx.csv", index=False)

In [ ]:
import pickle

with open('/content/drive/MyDrive/NLP_MASTER/news_archive.pkl', 'wb') as f:
  pickle.dump(news_archive, f)

with open('/content/drive/MyDrive/NLP_MASTER/news_archive.pkl', 'rb') as f:
  loaded_news_archive = pickle.load(f)

In [ ]:
len(loaded_news_archive)